Use the fact that LSBs should not be sorted

maloo opened the pull request:
You might not want to merge this as is in case someone tries to use the Radix sort for normal ulongs. But it shows that this solution is at least twice as fast as the others. Also wondering how much you would gain by hard coding a radix r=8 version ;)

maloo committed:
Use the fact that LSBs should not be sorted

mgravell commented:
any chance you could make this configurable via args, perhaps defaulting to `0` or `64`? I think it is actually a very valid addition. Re hard-coding to `const` - there's only one way to find out :) feel free to add in a hard-coded version (or I can later - will be AFK for most of the day)

maloo committed:
Parameterized Radix ulong sort group size

Also added a Radix 10 sort, since it aligns well with 40 bit key and is slightly faster than 5 * 8 radix sort

maloo commented:
I did a quick test with dedicated r=8. But the more constants I added the slower it got. I didn't have time to analyze or check generated IL/assembly. But feels like a compiler "bug". If a constant makes it slower, at least compiler should make it as fast by introducing a variable. You're welcome to have a go at it :)
Also added a Radix r=10 case to my CombinedIndex since it was faster than 8. My key is 40 bits, so it aligns well with both 8 and 10.
Goes down from 1900ms to 350±50ms om my laptop.

mgravell commented:
I've gone a slightly different way with this, using a `ulong keyMask = ulong.MaxValue` argument. This is more flexible IMO, and doesn't seem to suffer any perf. If any r-length mask is zero, it skips that group - and stops completely when the overall mask becomes exhausted. I also added a "everything is in the same group" check, which allows the second phase to be skipped.
